### ISSUE TITLE
Correct use of fit_generator, predict_generator, and evaluate_generator

### ISSUE THREAD

**Issue by tbagnoli** [keras-team/keras#11754]
I'm encountering weird behaviour when using `fit_generator`, `predict_generator`, and `evaluate_generator`, and I would like to ask the following questions, for which I could find no answer in the documentation:

1. Is it ok to have batches of different sizes when using `fit_generator`?

My batches are defined timewise: they group the events that took place in the same hour. Each batch can therefore group a different number of events. For clarity, this is how my generators look like (following the logic in [this thread](https://stackoverflow.com/questions/46647008/keras-aggregated-objective-function)):

```
def grouper(g,x,y):
     while True:
        for gr in g.unique():
            # this assigns indices to the entire set of values in g,
            # the subsects to all the rows in which g == gr
            indices = g == gr
            yield (x[indices],y[indices])

all_data_generator = grouper(df['batch_id'], X, Y)
train_generator = grouper(df.loc[df['set'] == 'train', 'batch_id'], X_train, Y_train)
validation_generator = grouper(df.loc[df['set'] == 'val', 'batch_id'], X_val, Y_val)
test_generator = grouper(df.loc[df['set'] == 'test', 'batch_id'], X_test, Y_test)
```

2. Is it ok to have different number of batches in `train_generator` and `validation_generator`? For clarity, I pass those two (different) numbers explicitly to `fit_generator` in the call:

```
train_batches = df.loc[df['set'] == 'train', 'batch_id'].nunique() # roughly 80% of batches
val_batches = df.loc[df['set'] == 'val', 'batch_id'].nunique() # roughly 20% of batches

history = fmodel.fit_generator(train_generator, 
                                 steps_per_epoch=train_batches, 
                                 validation_data=validation_generator,
                                 validation_steps=val_batches,
                                 epochs=20, verbose = 0)
```

3. Predictions are wildly different depending on whether I use `predict_classes` or `predict_generator`, which baffles me.

```
df['pred'] = fmodel.predict_classes(X)

# returns different results from
total_batches = df['batch_id'].nunique()
df['pred_gen'] = fmodel.predict_generator(data_generator, steps = total_batches)
```

4. Similarly, `evaluate` and `evaluate_generator` return different results:

```
scores = model.evaluate(X_test, Y_test, verbose = 0)

# returns different results from
scores_generator = fmodel.evaluate_generator(test_generator, steps=test_batches)
```

I know there are already many issues referring to my points 3. and 4. (e.g., [3477](https://github.com/keras-team/keras/issues/3477), [6499](https://github.com/keras-team/keras/issues/6499)) but the main takeaways there seem to refer to 
- using ImageDataGenerator with/without rescaling;
- shuffling data, which according to [fit_generator documentation](https://keras.io/models/sequential/) "Has no effect when steps_per_epoch is not None."
- using workers > 1, which by default is not the case.

So I'm wondering whether points 1. and 2. might be the culprits here.

**Comment by swierh** [comment]
> Predictions are wildly different depending on whether I use predict_classes or predict_generator, which baffles me.

When using `predict_generator` with multi-processing on (the default), it _uses a few more batches from the generator than the steps_. So the next time your generator is used, it doesn't start at the first sample, but at a few samples in (n_cores/2 on average).

Two potential fixes for this are calling `generator.reset()` before using it again, or setting `multi_processing=False`, example implementing both (you need only one):

```
generator.reset()
y_pred = model.predict_generator(generator, steps=n_steps, multi_processing=False)
```
